test/Container/ApplicationFactoryTest.php

```diff
@@ -1071,4 +1090,85 @@ public function testCanSpecifyRouteOptionsViaConfiguration()
         $this->assertInstanceOf(Route::class, $route);
         $this->assertEquals($expected, $route->getOptions());
     }
+
+    public function testCanPipeRouteSpecificMiddlewareViaConfiguration()
+    {
+        $router = $this->prophesize(RouterInterface::class);
+        $emitter = $this->prophesize(EmitterInterface::class);
+        $finalHandler = function ($req, $res, $err = null) {
+        };
+
+        $expected = [
+            'values' => [
+                'foo' => 'bar'
+            ],
+            'tokens' => [
+                'bar' => 'foo'
+            ]
+        ];
+        $config = [
+            'routes' => [
+                [
+                    'path' => '/',
+                    'middleware' => [
+                        'Hello',
+                        function () {
+                            return 'World';
+                        }
+                    ],
+                    'name' => 'home',
+                    'allowed_methods' => ['GET'],
+                    'options' => $expected
+                ],
+            ],
+        ];
+
+        $this->container
+            ->has(RouterInterface::class)
+            ->willReturn(true);
```

Comment:
@michaelmoussa Rebase against latest master, and then use `$this->injectServiceInContainer($this->container, <service name>, <service/value/mock>)` to inject services in the container. (You'll see examples of this in the test class already). This will simplify your tests, and reduce the white noise, making the test setup easier to understand. If you need help, let me know.